File: _static/code/career/career.py
from quantecon.distributions import BetaBinomial


class CareerWorkerProblem:
    """
    An instance of the class is an object with data on a particular
    problem of this type, including probabilites, discount factor and
    sample space for the variables.

    Parameters
    ----------
    β : scalar(float), optional(default=5.0)
        Discount factor
    B : scalar(float), optional(default=0.95)
        Upper bound of for both ϵ and θ
    N : scalar(int), optional(default=50)
        Number of possible realizations for both ϵ and θ
    F_a : scalar(int or float), optional(default=1)
        Parameter `a` from the career distribution
    F_b : scalar(int or float), optional(default=1)
        Parameter `b` from the career distribution
    G_a : scalar(int or float), optional(default=1)
        Parameter `a` from the job distribution
    G_b : scalar(int or float), optional(default=1)
        Parameter `b` from the job distribution

    Attributes
    ----------
    β, B, N : see Parameters
    θ : array_like(float, ndim=1)
        A grid of values from 0 to B
    ϵ : array_like(float, ndim=1)
        A grid of values from 0 to B
    F_probs : array_like(float, ndim=1)
        The probabilities of different values for F
    G_probs : array_like(float, ndim=1)
        The probabilities of different values for G
    F_mean : scalar(float)
        The mean of the distribution for F
    G_mean : scalar(float)
        The mean of the distribution for G

    """

    def __init__(self, B=5.0, β=0.95, N=50, F_a=1, F_b=1, G_a=1,
                 G_b=1):
        self.β, self.N, self.B = β, N, B
        self.θ = np.linspace(0, B, N)     # set of θ values
        self.ϵ = np.linspace(0, B, N)     # set of ϵ values
        self.F_probs = BetaBinomial(N-1, F_a, F_b).pdf()
        self.G_probs = BetaBinomial(N-1, G_a, G_b).pdf()
        self.F_mean = np.sum(self.θ * self.F_probs)
        self.G_mean = np.sum(self.ϵ * self.G_probs)

        # Store these parameters for str and repr methods
        self._F_a, self._F_b = F_a, F_b
        self._G_a, self._G_b = G_a, G_b


    def bellman_operator(self, v):
        """
        The Bellman operator for the career / job choice model of Neal.

        Parameters
        ----------
        v : array_like(float)
            A 2D NumPy array representing the value function
            Interpretation: :math:`v[i, j] = v(\theta_i, \epsilon_j)`

        Returns
        -------
        new_v : array_like(float)
            The updated value function Tv as an array of shape v.shape

        """
        new_v = np.empty(v.shape)
        for i in range(self.N):
            for j in range(self.N):
                # stay put
                v1 = self.θ[i] + self.ϵ[j] + self.β * v[i, j]

                # new job
                v2 = self.θ[i] + self.G_mean + self.β * v[i, :] @ self.G_probs

                # new life
                v3 = self.G_mean + self.F_mean + self.β * self.F_probs @ v @ self.G_probs
                new_v[i, j] = max(v1, v2, v3)
        return new_v

    def get_greedy(self, v):
        """
        Compute optimal actions taking v as the value function.

        Parameters
        ----------
        v : array_like(float)
            A 2D NumPy array representing the value function
            Interpretation: :math:`v[i, j] = v(\theta_i, \epsilon_j)`

        Returns
        -------
        policy : array_like(float)
            A 2D NumPy array, where policy[i, j] is the optimal action
            at :math:`(\theta_i, \epsilon_j)`.

            The optimal action is represented as an integer in the set
            1, 2, 3, where 1 = 'stay put', 2 = 'new job' and 3 = 'new
            life'

        """
        policy = np.empty(v.shape, dtype=int)
        for i in range(self.N):
            for j in range(self.N):
                v1 = self.θ[i] + self.ϵ[j] + self.β * v[i, j]
                v2 = self.θ[i] + self.G_mean + self.β * v[i, :] @ self.G_probs
                v3 = self.G_mean + self.F_mean + self.β * self.F_probs @ v @ self.G_probs
                if v1 > max(v2, v3):
                    action = 1
                elif v2 > max(v1, v3):
                    action = 2
                else:
                    action = 3
                policy[i, j] = action

        return policy

File: rst_files/_static/code/career/career.jl
"""
Career/job choice model of Derek Neal (1999)

##### Fields

- `β::AbstractFloat` : Discount factor in (0, 1)
- `N::Integer` : Number of possible realizations of both ϵ and θ
- `B::AbstractFloat` : upper bound for both ϵ and θ
- `θ::AbstractVector` : A grid of values on [0, B]
- `ϵ::AbstractVector` : A grid of values on [0, B]
- `F_probs::AbstractVector` : The pdf of each value associated with of F
- `G_probs::AbstractVector` : The pdf of each value associated with of G
- `F_mean::AbstractFloat` : The mean of the distribution F
- `G_mean::AbstractFloat` : The mean of the distribution G

"""

struct CareerWorkerProblem{TF<:AbstractFloat,
                           TI<:Integer,
                           TAV<:AbstractVector{TF},
                           TAV2<:AbstractVector{TF}}
    β::TF
    N::TI
    B::TF
    θ::TAV
    ϵ::TAV
    F_probs::TAV2
    G_probs::TAV2
    F_mean::TF
    G_mean::TF
end


"""
Constructor with default values for `CareerWorkerProblem`

##### Arguments

- `β::Real(0.95)` : Discount factor in (0, 1)
- `B::Real(5.0)` : upper bound for both ϵ and θ
- `N::Real(50)` : Number of possible realizations of both ϵ and θ
- `F_a::Real(1), F_b::Real(1)` : Parameters of the distribution F
- `G_a::Real(1), G_b::Real(1)` : Parameters of the distribution F

##### Notes

There is also a version of this function that accepts keyword arguments for
each parameter
"""
# use key word argument
function CareerWorkerProblem{TF<:AbstractFloat}(;β::TF=0.95,
                                                 B::TF=5.0,
                                                 N::Integer=50,
                                                 F_a::TF=1.0,
                                                 F_b::TF=1.0,
                                                 G_a::TF=1.0,
                                                 G_b::TF=1.0)
    θ = linspace(0, B, N)
    ϵ = copy(θ)
    dist_F = BetaBinomial(N-1, F_a, F_b)
    dist_G = BetaBinomial(N-1, G_a, G_b)
    F_probs = pdf.(dist_F, support(dist_F))
    G_probs = pdf.(dist_G, support(dist_G))
    F_mean = sum(θ .* F_probs)
    G_mean = sum(ϵ .* G_probs)
    CareerWorkerProblem(β, N, B, θ, ϵ, F_probs, G_probs, F_mean, G_mean)
end


"""
Apply the Bellman operator for a given model and initial value.

##### Arguments

- `cp::CareerWorkerProblem` : Instance of `CareerWorkerProblem`
- `v::Matrix`: Current guess for the value function
- `out::Matrix` : Storage for output
- `;ret_policy::Bool(false)`: Toggles return of value or policy functions

##### Returns

None, `out` is updated in place. If `ret_policy == true` out is filled with the
policy function, otherwise the value function is stored in `out`.

"""
function update_bellman!(cp::CareerWorkerProblem,
                         v::Array,
                         out::Array;
                         ret_policy::Bool=false)

    # new life. This is a function of the distribution parameters and is
    # always constant. No need to recompute it in the loop
    v3 = (cp.G_mean + cp.F_mean + cp.β .*
          cp.F_probs' * v * cp.G_probs)[1]        # don't need 1 element array

    for j=1:cp.N
        for i=1:cp.N
            # stay put
            v1 = cp.θ[i] + cp.ϵ[j] + cp.β * v[i, j]

            # new job
            v2 = (cp.θ[i] .+ cp.G_mean .+ cp.β .*
                  v[i, :]' * cp.G_probs)[1]       # don't need a single element array

            if ret_policy
                if v1 > max(v2, v3)
                    action = 1
                elseif v2 > max(v1, v3)
                    action = 2
                else
                    action = 3
                end
                out[i, j] = action
            else
                out[i, j] = max(v1, v2, v3)
            end
        end
    end
end


function update_bellman(cp::CareerWorkerProblem, v::Array; ret_policy::Bool=false)
    out = similar(v)
    update_bellman!(cp, v, out, ret_policy=ret_policy)
    return out
end

"""
Extract the greedy policy (policy function) of the model.

##### Arguments

- `cp::CareerWorkerProblem` : Instance of `CareerWorkerProblem`
- `v::Matrix`: Current guess for the value function
- `out::Matrix` : Storage for output

##### Returns

None, `out` is updated in place to hold the policy function

"""
function get_greedy!(cp::CareerWorkerProblem, v::Array, out::Array)
    update_bellman!(cp, v, out, ret_policy=true)
end


function get_greedy(cp::CareerWorkerProblem, v::Array)
    update_bellman(cp, v, ret_policy=true)
end
